api/xds_protocol.rst

```diff
@@ -469,3 +469,155 @@ occurred via a :ref:`resource update <xds_protocol_resource_update>`.
 
 .. |Multiple EDS requests on the same stream| image:: diagrams/eds-same-stream.svg
 .. |Multiple EDS requests on distinct streams| image:: diagrams/eds-distinct-stream.svg
+
+Implementations
+---------------
+
+.. _on_demand_vhds_rds_protocol:
+
+Virtual Host Discovery Service
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+By default in RDS, all routes for a cluster are sent to every Envoy instance
+in the mesh. This causes scaling issues as the size of the cluster grows. The
+majority of this complexity can be found in the virtual host configurations, of
+which most are not needed by any individual proxy. 
+
+In order to fix this issue, the Virtual Host Discovery Service (VHDS) protocol
+uses the delta xDS protocol to allow a route configuration to be subscribed to
+and the necessary virtual hosts to be requested as needed. Instead of sending
+all virtual hosts with a route config, using VHDS will allow an Envoy instance
+to subscribe and unsubscribe from a list of virtual hosts stored internally in
+the xDS management server. The xDS management server will monitor this list and
+use it to filter the configuration sent to an individual Envoy instance to only
+contain the subscribed virtual hosts.
+
+In VHDS, Envoy will send a
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+with the 
+:ref:`type_url <envoy_api_field_DeltaDiscoveryRequest.type_url>` set to 
+`type.googleapis.com/envoy.api.v2.route.VirtualHost` 
+and :ref:`resource_names_subscribe <envoy_api_field_DeltaDiscoveryRequest.resource_names_subscribe>` 
+set to a list of virtual host resource names for which it would like
+configuration. The management server will respond with a
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+with the :ref:`resources <envoy_api_field_DeltaDiscoveryResponse.resources>`
+field populated with the specified virtual hosts, the 
+:ref:`name <envoy_api_field_resource.name>` field populated with
+the virtual host name, and the 
```

Comment:
Shouldn't the `name` and `aliases` fields contain the strings of the form `<route configuration name>/<host entry>`, so that one of them exactly matches the requested resource name?  Otherwise, the client needs special logic to determine which of the returned resources matches which route configuration.

Reply:
VHDS on-demand requests do not contain resource names, as all we have at point in time is the contents of host header; this is what precipitated the need for aliases.

Consequently, when a reply comes in, virtual host name isn't useful to us, as we still don't know what domains it maps to. Hence the need to return aliases with the response.

Reply:
We have both the host header and the route configuration name.  If the VHDS resource has an alias for every domain it will need to match against, then there should be no problem having an exact match here.

Reply:
Yeah, I think we need to have the resources/aliases follow the new namespace convention for consistency and avoid special casing.

Reply:
oh, apologies, misunderstood the comment. Agreed, esp. re: aliases field.

Reply:
Clarified